### PR TITLE
10 backend setup

### LINKED ISSUE #10: Setup Backend
**Goal**
- Setup boilerplate for backend

**Tasks**
- [x]  Create file structure
  - Note: I propose we use a layered structure like here: https://nestjs.com/
- [x] Setup env variable configuration
  - include `sample.env` file to document env variable names
- [x] Setup logging
- [x] Setup Express app
  - Similar to Flask app
  - Register routes & middlewares
- [x] Setup middlewares
  - HTTP logging
  - Error handling
- [x] Setup routing
  - `/api/health`
- [x] Setup database
  - Mongoose connection

**Criteria**
- `npm run dev`
  - Starts up hot reloading
  - Suggested tests: 
    - edit `index.ts` to display message in console while app is running
    -  go to `/api/health` API endpoint and see server running status, database health, etc

- `npm run lint`:
  - performs linting and formatting on code
  - Suggessted tests: 
    - add an extra blank line at the end of `index.ts` and run `npm run lint` to see if it is removed

### PR DESCRIPTION
**Pull request by rezbee-dev**
Closes #10 

Instructions: 
- `git fetch remote pull/22/head:pr-22`
- `git checkout pr-22`
- `cd backend`
- `npm i`
- `npm run db:up`
- `npm run dev`

### CODE REVIEW

File: backend/package.json [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+  "scripts": {
+    "build": "rimraf ./build && tsc",
+    "start": "node build/index.js",
+    "dev": "cross-env NODE_ENV=dev ts-node-dev --respawn src/index.ts",
+    "test": "echo \"Error: no test specified\" && exit 1",
+    "lint": "eslint --ignore-path ../.gitignore --ext .ts . --fix",
+    "lint:check": "eslint --ignore-path ../.gitignore --ext .ts . --fix",
+    "format": "prettier --ignore-path ../.gitignore --write 'src/**/*.ts'",
+    "format:check": "prettier --ignore-path ../.gitignore --check 'src/**/*.ts'",
+    "db:up": "docker compose -f db.docker-compose.yml up -d",
```
Comment (monocle) [inline backend/package.json:15]: Once we have the backend in a container, I'm not sure you'll be running these docker compose commands here. Should they be removed?

File: backend/src/api/app.ts [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+const app = express();
+
+app.use(cors({ origin: `http://localhost:${env.FRONTEND_PORT}` }));
+app.use(express.json());
+if (process.env.NODE_ENV === "dev") app.use(httpLogger());
+app.use(router);
+app.use(errorHandler);
+app.use(unknownEndpoint);
+
+export default app;
```
Comment (monocle) [inline backend/src/api/app.ts:17]: I think it's common practice to have a blank line at the end of files.

File: backend/src/common/utils.ts [hunk 3]
```diff
@@ -0,0 +1,17 @@
+import { sep } from "path";
+
+// Todo: write tests for this
```
Comment (monocle) [inline backend/src/common/utils.ts:3]: Should we just create GH issues for any todos?

File: backend/src/data/db.ts [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+import { Logger } from '../common/logger'
+
+const start = async (): Promise<string> => {
+    try {
+        await mongoose.connect(env.getMongoUri())
+        return `Database connected at ${env.getMongoUri()}`;
+    } catch (error) {
+        Logger.error("Could not connect to database!")
+        if (env.NODE_ENV !== "production") {
+            Logger.error("Did you run 'npm run db:up'?")
```
Comment (monocle) [inline backend/src/data/db.ts:12]: Might need to change this error message if we don't use npm run to bring up the database.

File: backend/src/index.ts [hunk 5]
```diff
@@ -0,0 +1,26 @@
+import { env } from "./common/env";
+import startServer from "./api/server";
+import { Utils } from "./common/utils";
+import { Logger } from "./common/logger";
+import { db } from "./data/db";
+
+console.log("hello world");
```
Comment (monocle) [inline backend/src/index.ts:7]: Maybe remove console.logs with PRs?

**Review by monocle — COMMENTED** [review]
How about updating `.gitignore` then we'll merge this and do the server Dockerfile as a separate PR?

File: backend/package.json (hunk 1 above)
Comment (rezbee-dev) [reply backend/package.json:15]: This is just my personal preference to use docker for database stuff
I don't plan to use docker for the entire app for development

I think it shouldn't affect things 🤞

File: backend/src/api/app.ts (hunk 2 above)
Comment (rezbee-dev) [reply backend/src/api/app.ts:17]: Sorry, I forgot to run linters & formatters on the code. I don't think my vscode is setup properly.... I will make the fix

I think once we get the branch protections setup, these problems should be automatically flagged

File: backend/src/common/utils.ts (hunk 3 above)
Comment (rezbee-dev) [reply backend/src/common/utils.ts:3]: Sorry these were just personal comments I made. In the future, I'll make them into git issues and delete the comments

File: backend/src/data/db.ts (hunk 4 above)
Comment (rezbee-dev) [reply backend/src/data/db.ts:12]: Ok will do

File: backend/src/index.ts (hunk 5 above)
Comment (rezbee-dev) [reply backend/src/index.ts:7]: What's PRs?

But yea, I will remove the console.logs. I forgot to do that. In fact, I think we can add a ESLint rule to automatically flag `console.logs` in the code (for backend)

Comment (rezbee-dev) [reply backend/src/index.ts]: https://github.com/EastWestCollab/samba/pull/23/commits/a78943034a9b203e90ff6a1131766a652414c15c

File: backend/src/data/db.ts (hunk 4 above)
Comment (rezbee-dev) [reply backend/src/data/db.ts]: https://github.com/EastWestCollab/samba/pull/23/commits/d5b097350a6f1c3b92db62ea5a242c475a9b8776

File: backend/src/common/utils.ts (hunk 3 above)
Comment (rezbee-dev) [reply backend/src/common/utils.ts]: https://github.com/EastWestCollab/samba/pull/23/commits/5a870260f0826083786eca77602094d4db9d7cf1

File: backend/src/api/app.ts (hunk 2 above)
Comment (rezbee-dev) [reply backend/src/api/app.ts]: https://github.com/EastWestCollab/samba/pull/23/commits/decacfee1c1286e2c15317f1f3ec9d093054b664

**Review by monocle — APPROVED** [review]

**Pull request merged**